compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/cyclefind/GlobalVariableRefAnalyzer.java

```diff
@@ -0,0 +1,355 @@
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Analyze global variable reference patterns and reorder them in reverse dependency order.
+ */
+public class GlobalVariableRefAnalyzer {
+    private final BLangDiagnosticLog dlog;
+    private final BLangPackage pkgNode;
+    private final Map<BSymbol, List<BSymbol>> globalNodeDependsOn;
+    private final Map<BSymbol, NodeInfo> dependencyNodes;
+    private final Deque<NodeInfo> nodeInfoStack;
+    private final List<List<NodeInfo>> cycles;
+    private final List<NodeInfo> dependencyOrder;
+    private int curNodeId;
+
+    public GlobalVariableRefAnalyzer(BLangPackage pkgNode,
+                                     BLangDiagnosticLog dlog,
+                                     Map<BSymbol, List<BSymbol>> globalNodeDependsOn) {
+        this.pkgNode = pkgNode;
+        this.dlog = dlog;
+        this.globalNodeDependsOn = globalNodeDependsOn;
+        this.dependencyNodes = new HashMap<>();
+        this.cycles = new ArrayList<>();
+        this.nodeInfoStack = new ArrayDeque<>();
+        this.dependencyOrder = new ArrayList<>();
+    }
+
+    /**
+     * Analyze the global variable references and reorder them or emit error if they contain cyclic references.
+     */
+    public void analyzeAndReOrder() {
+        List<BSymbol> globalVarsAndDependentFuncs = getGlobalVariablesAndDependentFunctions();
+
+        pruneDependencyRelations(this.globalNodeDependsOn);
+
+        Set<BSymbol> sorted = new LinkedHashSet<>();
+        LinkedList<BSymbol> dependencies = new LinkedList<>();
+
+        for (BSymbol symbol : globalVarsAndDependentFuncs) {
+            // Only analyze unvisited nodes.
+            // Do DFS into dependency providers to detect cycles.
+            if (!dependencyNodes.containsKey(symbol)) {
+                NodeInfo node = new NodeInfo(curNodeId++, symbol);
+                dependencyNodes.put(symbol, node);
+                analyzeProvidersRecursively(node);
+            }
+            // Extract all the dependencies found in last call to analyzeProvidersRecursively
+            if (!dependencyOrder.isEmpty()) {
+                List<BSymbol> symbolsProvidersOrdered = dependencyOrder.stream()
+                        .map(nodeInfo -> nodeInfo.symbol)
+                        .collect(Collectors.toList());
+                dependencies.addAll(symbolsProvidersOrdered);
+                dependencyOrder.clear();
+            }
+
+            List<BSymbol> symbolsProviders = globalNodeDependsOn.get(symbol);
+            boolean symbolHasProviders = symbolsProviders != null && !symbolsProviders.isEmpty();
+
+            // Independent variable declaration, add to sorted list.
+            if (!symbolHasProviders && (!sorted.contains(symbol))) {
```

Comment:
both if conditions(106 and 110) evaluates !sorted.contains(symbol) so can we use single if condition for both checks?